bin/run

```diff
@@ -18,8 +20,11 @@ number_validator = NumberValidator.new
 prompter = Prompt.new(console:, number_validator:,
                       board:)
 ui = UI.new(display:, prompter:)
-player_one = Player.new(marker: 'X')
-player_two = Player.new(marker: 'O')
+random = Random.new(ENV['SEED'].to_i || Random.new_seed)
+rand = ->(range) { random.rand(range) }
+random_number_generator = RandomNumberGenerator.new(range: 1..9, rand:)
```

Comment:
I have to say I disagree with this setup, it means that a user can disable randomness itself by setting the environment variable. I feel this new abstraction layer surfaced by the need to create unit tests and so it doesn't feel clean.

Using unit tests for random functions may not the best approach (see [this answer](https://softwareengineering.stackexchange.com/a/147142)), as randomness is a behaviour, not a repeatable pattern.

We can do something similar to how this gem tests the race condition problem:

https://github.com/EByrdS/szymanskis_mutex/blob/ecb1c1b5dcb3f7ed4cbc7d093d7c9acfb3eb5199/spec/szymanskis_mutex_spec.rb#L31-L42

We can build a couple of tests to check for general patterns when the computer chooses a random move, and we should be able to set a seed when testing more integration-like cases, inside the test itself.

Reply:
I’m excited to discuss the pros and cons of different approaches to dealing with randomness in a program! A good amount of this discussion is over my head at the moment since this is the first time I’ve tried to test a program with random behavior, so I’d like to make sure I'm understanding and get some more information.   

I hear you saying that you see the cons of the current approach as introducing a new abstraction layer for the sake of unit testing, and giving the user the ability to configure randomness.

  I believe you’re suggesting that, instead of having unit tests that use mocks to determine whether the program is correctly interacting with Ruby’s Random library, the unit tests would look at the actual random output and assert that it’s following a pattern. Then, the integration-level (and higher?) tests would set a seed within the spec file itself to get predictable output.  
  
  If I’m understanding correctly, this would make it so that the computer player’s behavior was always random except within the higher-level tests, and you’re saying that the ‘pro’ of this approach would be that only the programmer would be able to change the random behavior.   
  
  Do you see any tradeoffs to switching approaches?